backends/dnf/pk-backend-dnf.c

```diff
@@ -1841,6 +1841,197 @@ dnf_utils_find_package_ids (DnfSack *sack, gchar **package_ids, GError **error)
 	return hash;
 }
 
+static void
+dnf_utils_process_dependency (DnfSack *sack,
+			      DnfPackage *pkg,
+			      PkRoleEnum role,
+			      gboolean recursive,
+			      GHashTable *visited,
+			      GPtrArray *results_array,
+			      GError **error)
+{
+	g_autoptr(GQueue) queue = g_queue_new ();
+	g_queue_push_tail (queue, g_object_ref (pkg));
+
+	/* mark the start package as visited so we don't process it again */
+	if (!g_hash_table_contains (visited, dnf_package_get_package_id (pkg)))
+		g_hash_table_add (visited, g_strdup (dnf_package_get_package_id (pkg)));
+
+	while (!g_queue_is_empty (queue)) {
+		g_autoptr(DnfPackage) curr = g_queue_pop_head (queue);
+		g_autoptr(DnfReldepList) reldeps = NULL;
+		guint n;
+
+		if (role == PK_ROLE_ENUM_DEPENDS_ON)
+			reldeps = dnf_package_get_requires (curr);
+		else
+			reldeps = dnf_package_get_provides (curr);
+
+		n = dnf_reldep_list_count (reldeps);
+		for (guint i = 0; i < n; i++) {
+			DnfReldep *reldep = dnf_reldep_list_index (reldeps, i);
+			const gchar *req = dnf_reldep_to_string (reldep);
+			hy_autoquery HyQuery query = hy_query_create (sack);
+			GPtrArray *results;
+
+			/* find packages that provide the requirement, or require the provided capability */
+			if (role == PK_ROLE_ENUM_DEPENDS_ON)
+				hy_query_filter (query, HY_PKG_PROVIDES, HY_EQ, req);
+			else
+				hy_query_filter (query, HY_PKG_REQUIRES, HY_EQ, req);
+
+			results = dnf_utils_run_query_with_newest_filter (sack, query);
+			for (guint j = 0; j < results->len; j++) {
+				DnfPackage *res = g_ptr_array_index (results, j);
+				const gchar *resid = dnf_package_get_package_id (res);
+
+				/* ignore self-referential dependencies */
+				if (g_strcmp0 (resid, dnf_package_get_package_id (curr)) == 0)
+					continue;
+
+				if (!g_hash_table_contains (visited, resid)) {
+					g_hash_table_add (visited, g_strdup (resid));
+					g_ptr_array_add (results_array, g_object_ref (res));
+					if (recursive)
+						g_queue_push_tail (queue, g_object_ref (res));
+				}
+			}
+			g_ptr_array_unref (results);
```

Comment:
```suggestion

```

Reply:
Oh I see, combined with the thing above.